File: tests/http_python/test_http.py
"""HTTP library test script for malwi integration tests.

Each function makes one HTTP request using a specific library.
The target URL is passed as the first argument (typically a local test server).

Usage: python test_http.py <url> <library>
  library: stdlib | requests | httpx | aiohttp | urllib3
"""
import sys


def test_stdlib(url):
    """Test http.client (standard library)."""
    import http.client
    import urllib.parse

    parsed = urllib.parse.urlparse(url)
    host = parsed.hostname
    port = parsed.port or (443 if parsed.scheme == "https" else 80)

    if parsed.scheme == "https":
        conn = http.client.HTTPSConnection(host, port)
    else:
        conn = http.client.HTTPConnection(host, port)

    conn.request("GET", parsed.path or "/")
    response = conn.getresponse()
    response.read()
    conn.close()
    print(f"stdlib: {response.status}")


def test_urllib(url):
    """Test urllib.request (standard library)."""
    import urllib.request
    resp = urllib.request.urlopen(url)
    resp.read()
    print(f"urllib: {resp.status}")


def test_requests(url):
    """Test requests library."""
    import requests
    resp = requests.get(url)
    print(f"requests: {resp.status_code}")


def test_requests_post(url):
    """Test requests.post with JSON body."""
    import requests
    resp = requests.post(url, json={"key": "value"})
    print(f"requests.post: {resp.status_code}")


def test_httpx(url):
    """Test httpx library."""
    import httpx
    resp = httpx.get(url)
    print(f"httpx: {resp.status_code}")


def test_aiohttp(url):
    """Test aiohttp library."""
    import aiohttp
    import asyncio

    async def fetch():
        async with aiohttp.ClientSession() as session:
            async with session.get(url) as resp:
                await resp.text()
                print(f"aiohttp: {resp.status}")

    asyncio.run(fetch())


def test_urllib3(url):
    """Test urllib3 library."""
    import urllib3
    pool = urllib3.PoolManager()
    resp = pool.request("GET", url)
    print(f"urllib3: {resp.status}")


if __name__ == "__main__":
    if len(sys.argv) < 3:
        print(f"Usage: {sys.argv[0]} <url> <library>")
        sys.exit(1)

    url = sys.argv[1]
    library = sys.argv[2]

    dispatch = {
        "stdlib": test_stdlib,
        "urllib": test_urllib,
        "requests": test_requests,
        "requests_post": test_requests_post,
        "httpx": test_httpx,
        "aiohttp": test_aiohttp,
        "urllib3": test_urllib3,
    }

    if library not in dispatch:
        print(f"Unknown library: {library}")
        sys.exit(1)

    dispatch[library](url)

File: tests/http_node/test_http.js
/**
 * HTTP library test script for malwi integration tests.
 *
 * Each function makes one HTTP request using a specific library.
 * The target URL is passed as the first argument.
 *
 * Usage: node test_http.js <url> <library>
 *   library: http | https | axios | got | node-fetch
 */

const url = process.argv[2];
const library = process.argv[3];

if (!url || !library) {
  console.log('Usage: node test_http.js <url> <library>');
  process.exit(1);
}

function testHttp(targetUrl) {
  return new Promise((resolve, reject) => {
    const http = require('http');
    http.get(targetUrl, (res) => {
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => {
        console.log('http:', res.statusCode);
        resolve();
      });
    }).on('error', reject);
  });
}

function testHttps(targetUrl) {
  return new Promise((resolve, reject) => {
    const https = require('https');
    https.get(targetUrl, (res) => {
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => {
        console.log('https:', res.statusCode);
        resolve();
      });
    }).on('error', reject);
  });
}

async function testAxios(targetUrl) {
  const axios = require('axios');
  const resp = await axios.get(targetUrl);
  console.log('axios:', resp.status);
}

async function testGot(targetUrl) {
  const got = require('got');
  const resp = await got(targetUrl);
  console.log('got:', resp.statusCode);
}

async function testNodeFetch(targetUrl) {
  const fetch = require('node-fetch');
  const resp = await fetch(targetUrl);
  console.log('node-fetch:', resp.status);
}

const dispatch = {
  'http': testHttp,
  'https': testHttps,
  'axios': testAxios,
  'got': testGot,
  'node-fetch': testNodeFetch,
};

const fn = dispatch[library];
if (!fn) {
  console.log('Unknown library:', library);
  process.exit(1);
}

fn(url).catch((err) => {
  console.error('Error:', err.message);
  process.exit(1);
});
